### ISSUE TITLE
Fail to garbage collect Pandas dataframes

### ISSUE THREAD

**Issue by mrocklin** [dask/distributed#956]
Sometimes we fail to clean up Pandas dataframes.  Here are a couple of examples using dask.dataframe and dask.array on a "cluster" in the local environment.  In the dask.array case we cleanly remove used data.  In the dask.dataframe case the memory of our process remains high even after we delete things.

### Clean case with dask.array

```python
In [1]: import psutil

In [2]: proc = psutil.Process()

In [3]: proc.memory_info().rss / 1e6
Out[3]: 39.768064

In [4]: from distributed import Client, LocalCluster

In [5]: cluster = LocalCluster(nanny=False, n_workers=4, threads_per_worker=1)

In [6]: client = Client(cluster)

In [7]: bokeh.command.util - CRITICAL - Cannot start Bokeh server, port 8787 is already in use
In [7]: 

In [7]: import dask.array as da

In [8]: x = da.random.random(100000000, chunks=(1000000,)).persist()

In [9]: proc.memory_info().rss / 1e6
Out[9]: 872.411136

In [10]: proc.memory_info().rss / 1e6
Out[10]: 872.411136

In [11]: proc.memory_info().rss / 1e6
Out[11]: 872.411136

In [12]: del x

In [13]: proc.memory_info().rss / 1e6
Out[13]: 72.081408

In [14]: proc.memory_info().rss / 1e6
Out[14]: 72.081408
```

### Leaking case with Pandas dataframes

```python
In [1]: import psutil

In [2]: proc = psutil.Process()

In [3]: proc.memory_info().rss / 1e6
Out[3]: 38.109184

In [4]: proc.memory_info().rss / 1e6
Out[4]: 38.37952

In [5]: from distributed import Client, LocalCluster

In [6]: proc.memory_info().rss / 1e6
Out[6]: 51.851264

In [7]: cluster = LocalCluster(nanny=False, n_workers=4, threads_per_worker=1)

In [8]: client = Client(cluster)bokeh.command.util - CRITICAL - Cannot start Boe
In [8]: client = Client(cluster)

In [9]: proc.memory_info().rss / 1e6
Out[9]: 65.073152

In [10]: import dask.dataframe as dd

In [11]: df = dd.demo.daily_stock('GOOG', '2008', '2010', freq='1s', 
    ...:                          random_state=1234).persist()

In [12]: proc.memory_info().rss / 1e6
Out[12]: 584.060928

In [13]: proc.memory_info().rss / 1e6
Out[13]: 685.867008

In [14]: proc.memory_info().rss / 1e6
Out[14]: 804.74112

In [15]: proc.memory_info().rss / 1e6
Out[15]: 927.670272

In [16]: proc.memory_info().rss / 1e6
Out[16]: 996.499456

In [17]: proc.memory_info().rss / 1e6
Out[17]: 997.445632

In [18]: proc.memory_info().rss / 1e6
Out[18]: 997.445632

In [19]: del df

In [20]: cluster.scheduler.tasks
Out[20]: {}

In [21]: proc.memory_info().rss / 1e6
Out[21]: 997.711872
```

cc @pitrou and @jreback any thoughts on how to track this down or avoid it?

**Comment by mrocklin** [comment]
@jreback are there any known problems with memory leaks in Pandas?

@pitrou do you have any thoughts on how to identify where this leak occurs?

**Comment by pitrou** [comment]
A word of warning: `rss` tells you the size currently allocated by the kernel, but this could be memory which has been freed by the application while being held by the libc (a quick look at Pandas suggests it uses C `malloc()`). Try allocating a new dataframe of the same size and see whether memory size grows or not.

As a sidenote, `dd.demo.daily_stock` fails here. What is the `pandas_datareader` module?
```
>>> df = dd.demo.daily_stock('GOOG', '2008', '2010', freq='1s', random_state=1234).persist()
Traceback (most recent call last):
  File "<ipython-input-21-f70de842f2db>", line 1, in <module>
    df = dd.demo.daily_stock('GOOG', '2008', '2010', freq='1s', random_state=1234).persist()
  File "/home/antoine/dask/dask/dataframe/io/demo.py", line 186, in daily_stock
    from pandas_datareader import data
ModuleNotFoundError: No module named 'pandas_datareader'
```

**Comment by jreback** [comment]
```
In [1]: import psutil

In [2]: proc = psutil.Process()

In [3]: proc.memory_info().rss / 1e6
Out[3]: 74.919936

In [4]: proc.memory_info().rss / 1e6
Out[4]: 74.981376

In [5]: df = pd.DataFrame(np.random.randn(100000, 10))

In [6]: proc.memory_info().rss / 1e6
Out[6]: 83.06688

In [7]: proc.memory_info().rss / 1e6
Out[7]: 83.070976

In [8]: del df

In [9]: proc.memory_info().rss / 1e6
Out[9]: 83.095552

In [10]: df = pd.DataFrame(np.random.randn(100000, 10))

In [11]: proc.memory_info().rss / 1e6
Out[11]: 83.140608
```

```
In [1]: import psutil

In [2]: proc = psutil.Process()

In [3]: proc.memory_info().rss / 1e6
Out[3]: 74.661888

In [4]: arr = np.random.randn(100000, 10)

In [5]: proc.memory_info().rss / 1e6
Out[5]: 82.767872

In [6]: del arr

In [7]: proc.memory_info().rss / 1e6
Out[7]: 82.788352

In [8]: arr = np.random.randn(100000, 10)

In [9]: proc.memory_info().rss / 1e6
Out[9]: 82.849792

In [10]: del arr

In [11]: proc.memory_info().rss / 1e6
Out[11]: 82.870272
```

**Comment by jreback** [comment]
you need to install ``pandas-datareader`` (pip/conda)

**Comment by pitrou** [comment]
Thanks @jreback. Here it seems that the Dask dataframe sometimes doesn't get deleted at all:
```
>>> df = dd.demo.daily_stock('GOOG', '2008', '2010', freq='1h', random_state=1234).persist()
>>> import sys, weakref, gc
>>> sys.getrefcount(df)
2     # <- this is the normal refcount for a local object
>>> df

Dask DataFrame Structure:
                       close     high      low     open
npartitions=505                                        
2008-01-02 09:00:00  float64  float64  float64  float64
2008-01-03 09:00:00      ...      ...      ...      ...
...                      ...      ...      ...      ...
2009-12-31 09:00:00      ...      ...      ...      ...
2009-12-31 16:00:00      ...      ...      ...      ...
Dask Name: from-delayed, 505 tasks
>>> wr = weakref.ref(df)
>>> del df   # <- this should delete it
>>> wr()

Dask DataFrame Structure:
                       close     high      low     open
npartitions=505                                        
2008-01-02 09:00:00  float64  float64  float64  float64
2008-01-03 09:00:00      ...      ...      ...      ...
...                      ...      ...      ...      ...
2009-12-31 09:00:00      ...      ...      ...      ...
2009-12-31 16:00:00      ...      ...      ...      ...
Dask Name: from-delayed, 505 tasks
>>> sys.getrefcount(wr())
13   # <- not only it wasn't deleted,  but the refcount was increased?!
>>> len(cluster.scheduler.who_has)
505   # <- all partitions are still there?

```

**Comment by pitrou** [comment]
Apparently printing a Dask dataframe under IPython could be doing weird things as well:
```
>>> sys.getrefcount(df)
2
>>> df

Dask DataFrame Structure:
                       close     high      low     open
npartitions=505                                        
2008-01-02 09:00:00  float64  float64  float64  float64
2008-01-03 09:00:00      ...      ...      ...      ...
...                      ...      ...      ...      ...
2009-12-31 09:00:00      ...      ...      ...      ...
2009-12-31 16:00:00      ...      ...      ...      ...
Dask Name: from-delayed, 505 tasks
>>> sys.getrefcount(df)
8
>>> len(cluster.scheduler.tasks)
1010
>>> len(cluster.scheduler.who_has)
505
>>> df.__repr__()
'Dask DataFrame Structure:\n                       close     high      low     open\nnpartitions=505                                        \n2008-01-02 09:00:00  float64  float64  float64  float64\n2008-01-03 09:00:00      ...      ...      ...      ...\n...                      ...      ...      ...      ...\n2009-12-31 09:00:00      ...      ...      ...      ...\n2009-12-31 16:00:00      ...      ...      ...      ...\nDask Name: from-delayed, 505 tasks'
>>> sys.getrefcount(df)
5
>>> df.__repr__()
'Dask DataFrame Structure:\n                       close     high      low     open\nnpartitions=505                                        \n2008-01-02 09:00:00  float64  float64  float64  float64\n2008-01-03 09:00:00      ...      ...      ...      ...\n...                      ...      ...      ...      ...\n2009-12-31 09:00:00      ...      ...      ...      ...\n2009-12-31 16:00:00      ...      ...      ...      ...\nDask Name: from-delayed, 505 tasks'
>>> sys.getrefcount(df)
5
>>> df

Dask DataFrame Structure:
                       close     high      low     open
npartitions=505                                        
2008-01-02 09:00:00  float64  float64  float64  float64
2008-01-03 09:00:00      ...      ...      ...      ...
...                      ...      ...      ...      ...
2009-12-31 09:00:00      ...      ...      ...      ...
2009-12-31 16:00:00      ...      ...      ...      ...
Dask Name: from-delayed, 505 tasks
>>> sys.getrefcount(df)
11
```

**Comment by pitrou** [comment]
Scratching the two last messages -- apparently this was because of IPython, not because of the Dask dataframe type (could reproduce with simply `object()`).

**Comment by pitrou** [comment]
Note that `distributed` is out of the picture. With no distributed client running:
```
>>> import dask.dataframe as dd
>>> df = dd.demo.daily_stock('GOOG', '2008', '2010', freq='1s', random_state=1235).persist()
>>> import psutil
>>> import weakref, sys, gc
>>> wr = weakref.ref(df)
>>> sys.getrefcount(df)
2
>>> del df
>>> wr() is None
True
>>> psutil.Process().memory_info().rss / 1e6
991.62112
```

**Comment by pitrou** [comment]
So it seems that deleting a Dask dataframe can leave a small number of underlying dataframes around (3 here):
```
>>> import dask.dataframe as dd
>>> import pandas as pd
>>> df = dd.demo.daily_stock('GOOG', '2008', '2010', freq='1h', random_state=1235).persist()
>>> 
>>> len(df.dask)
505
>>> len([x for x in df.dask.values() if isinstance(x, pd.DataFrame)])
505
>>> frames = [x for x in df.dask.values() if isinstance(x, pd.DataFrame)]
>>> len(frames)
505
>>> import weakref
>>> frames = [weakref.ref(x) for x in df.dask.values() if isinstance(x, pd.DataFrame)]
>>> len(frames)
505
>>> len([f for f in frames if f() is not None])
505
>>> wr = weakref.ref(df)
>>> del df
>>> wr() is not None
False
>>> len([f for f in frames if f() is not None])
3
```

Let's try to see what those frames are. They seem a bit random:
```
>>> a, b, c = [f() for f in frames if f() is not None]
>>> a.head()

                       close     high      low     open
timestamp                                              
2008-05-15 09:00:00  577.799  579.489  576.904  579.001
2008-05-15 10:00:00  579.922  579.922  575.611  575.611
2008-05-15 11:00:00  582.927  582.951  581.008  581.008
2008-05-15 12:00:00  582.438  582.438  581.623  582.312
2008-05-15 13:00:00  581.126  582.190  581.126  581.651
>>> b.head()

                       close     high      low     open
timestamp                                              
2008-05-22 09:00:00  554.211  554.211  551.951  551.951
2008-05-22 10:00:00  553.099  553.654  552.542  553.654
2008-05-22 11:00:00  552.344  554.152  552.344  554.152
2008-05-22 12:00:00  553.075  553.973  552.619  553.194
2008-05-22 13:00:00  546.648  552.450  546.648  552.450
>>> c.head()

                       close     high      low     open
timestamp                                              
2008-05-23 09:00:00  545.999  548.058  545.427  546.961
2008-05-23 10:00:00  546.713  546.713  541.679  545.211
2008-05-23 11:00:00  546.575  546.575  545.145  545.145
2008-05-23 12:00:00  552.030  552.030  547.238  548.377
2008-05-23 13:00:00  548.420  553.001  548.420  550.993
```

What is referring to the first dangling frame:
```
>>> [(type(r), r) for r in gc.get_referrers(a) if r is not globals()]

[(tuple, (                       close     high      low     open
   timestamp                                              
   2008-05-15 09:00:00  577.799  579.489  576.904  579.001
   2008-05-15 10:00:00  579.922  579.922  575.611  575.611
   2008-05-15 11:00:00  582.927  582.951  581.008  581.008
   2008-05-15 12:00:00  582.438  582.438  581.623  582.312
   2008-05-15 13:00:00  581.126  582.190  581.126  581.651
   2008-05-15 14:00:00  581.065  581.417  580.007  581.417
   2008-05-15 15:00:00  580.856  581.015  579.504  580.427
   2008-05-15 16:00:00  581.001  581.001  581.001  581.001,
   None,
   140221674002176)),
 (dict,
  {'name': 'loc',
   'ndim': 2,
   'obj':                        close     high      low     open
   timestamp                                              
   2008-05-15 09:00:00  577.799  579.489  576.904  579.001
   2008-05-15 10:00:00  579.922  579.922  575.611  575.611
   2008-05-15 11:00:00  582.927  582.951  581.008  581.008
   2008-05-15 12:00:00  582.438  582.438  581.623  582.312
   2008-05-15 13:00:00  581.126  582.190  581.126  581.651
   2008-05-15 14:00:00  581.065  581.417  580.007  581.417
   2008-05-15 15:00:00  580.856  581.015  579.504  580.427
   2008-05-15 16:00:00  581.001  581.001  581.001  581.001}),
 (dict,
  {'name': 'iloc',
   'ndim': 2,
   'obj':                        close     high      low     open
   timestamp                                              
   2008-05-15 09:00:00  577.799  579.489  576.904  579.001
   2008-05-15 10:00:00  579.922  579.922  575.611  575.611
   2008-05-15 11:00:00  582.927  582.951  581.008  581.008
   2008-05-15 12:00:00  582.438  582.438  581.623  582.312
   2008-05-15 13:00:00  581.126  582.190  581.126  581.651
   2008-05-15 14:00:00  581.065  581.417  580.007  581.417
   2008-05-15 15:00:00  580.856  581.015  579.504  580.427
   2008-05-15 16:00:00  581.001  581.001  581.001  581.001})]
```
Similar results can be had for the other dangling frames: they are referred to by a `(frame, None, integer timestamp)` tuple and two `{'name': 'loc' or 'iloc', 'ndim': 2, 'obj': frame}` dict.

At this point I don't know where to go next, as I don't know anything about Pandas internals. The `loc` and `iloc` things seem to suggest there may be some caching going on?

**Comment by jreback** [comment]
``.iloc/.loc`` are the pandas indexers, they keep a reference to the object (I don't think you need to explicity break this cycle as the gc would normally), but maybe here you do.

**Comment by jreback** [comment]
```
In [1]: df = pd.DataFrame(np.random.randn(10,2))

In [2]: df.iloc[0:5]
Out[2]: 
          0         1
0  0.060184  1.252374
1 -0.017412  0.741759
2  0.462292  0.092218
3  0.064403  0.790817
4  1.481308  0.962824

In [3]: df.iloc
Out[3]: <pandas.core.indexing._iLocIndexer at 0x10e748b70>

In [4]: df.iloc.obj
Out[4]: 
          0         1
0  0.060184  1.252374
1 -0.017412  0.741759
2  0.462292  0.092218
3  0.064403  0.790817
4  1.481308  0.962824
5 -0.826619  0.980654
6 -1.133114 -0.504807
7 -0.017072 -1.646783
8  0.928021 -0.175613
9 -0.332223  2.219120
```

**Comment by pitrou** [comment]
Oh, so you mean one needs to call `gc.collect()` (or wait for the GC to run automatically) in order for a Pandas dataframe to be *actually* collected:
```
>>> df = pd.DataFrame(np.random.randn(10,2))
>>> df.head()

          0         1
0 -0.390915 -0.197607
1 -1.096394 -0.347774
2 -0.724065  1.348458
3 -0.803112  1.071055
4  0.116906 -1.077839
>>> df.loc[:5].head()

          0         1
0 -0.390915 -0.197607
1 -1.096394 -0.347774
2 -0.724065  1.348458
3 -0.803112  1.071055
4  0.116906 -1.077839
>>> df.iloc[:5].head()

          0         1
0 -0.390915 -0.197607
1 -1.096394 -0.347774
2 -0.724065  1.348458
3 -0.803112  1.071055
4  0.116906 -1.077839
>>> import gc, sys, weakref
>>> sys.getrefcount(df)
4
>>> wr = weakref.ref(df)
>>> del df
>>> wr() is None
False
>>> gc.collect()
47
>>> wr() is None
True
```

I understand why some workers may make it look like memory hasn't been reclaimed after "deleting" a Dask dataframe and its underlying dataframes: all the underlying dataframes are not necessarily reclaimed up front, and depending on the allocator's heuristics, no memory at all may be returned to the OS.

**Comment by jreback** [comment]
> Oh, so you mean one needs to call gc.collect() (or wait for the GC to run automatically) in order for a Pandas dataframe to be actually collected:

those are the semantics of a cyclical ref, so yes. There are indexing allocations that are potentially done (e.g. not in all circumstances are these required), that may allocate c-structures. These have there own ``__dealloc__`` routines. so a collection is necessary. So you cannot guarantee *when* these are deleted. (again semantics of ``__del__``).

**Comment by pitrou** [comment]
> those are the semantics of a cyclical ref, so yes

Wouldn't it be better to only keep a weak reference? Users don't necessarily expect allocated data to linger around, especially for large objects such as dataframes.

**Comment by jreback** [comment]
yes it would

(this was done a long time ago though)

if u want to do a PR would be welcome!

**Comment by pitrou** [comment]
Regardless, even with `gc.collect()`, some frames are still lying around. They seem to hang on some Dask structures.

**Comment by jreback** [comment]
I created an issue on the pandas side. But I don't think this particular issue a pandas issue per-se, rather something in dask is holding onto a reference (and thus the object is not actually completely deleted).

**Comment by pitrou** [comment]
Ok, so in the pure Dask case (without Distributed), it is an issue with `multiprocessing.Pool` not releasing the references to the latest task(s). Distributed uses `concurrent.futures`, so a separate investigation will be needed.

**Comment by pitrou** [comment]
`multiprocessing` bug reported upstream at http://bugs.python.org/issue29861

**Comment by pitrou** [comment]
Ok, so in the Distributed case, this is an issue inside Distributed (in addition to the Pandas-specific `iloc` issue). I will submit a PR soon.

**Comment by jreback** [comment]
> (in addition to the Pandas-specific iloc issue)

this is not the problem per-se. really just making indexers more gc friendly.

**Comment by shughes-uk** [comment]
Seems to be happening again

```
In [1]: import psutil
   ...: from distributed import Client, LocalCluster
   ...: import dask.dataframe as dd
   ...: proc = psutil.Process()
   ...: 

In [2]: proc.memory_info().rss / 1e6
Out[2]: 114.638848

In [3]: cluster = LocalCluster(n_workers=4, threads_per_worker=1)

In [4]: client = Client(cluster)

In [5]: proc.memory_info().rss / 1e6
Out[5]: 125.964288

In [6]: df = dd.demo.daily_stock('GOOG', '2008', '2010', data_source='google',freq='1s', random_state=1234).persist()

In [7]: proc.memory_info().rss / 1e6
Out[7]: 153.74336

In [8]: proc.memory_info().rss / 1e6
Out[8]: 153.76384

In [9]: del df
   ...: 

In [10]: proc.memory_info().rss / 1e6
Out[10]: 154.312704

In [11]: cluster.scheduler.tasks
Out[11]: {}
```

Version info:
Python 3.5.2 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:52:12) 

```
In [4]: dask.__version__
Out[4]: '0.15.0'

In [5]: distributed.__version__
Out[5]: '1.17.0+1.gfb0b8f4'

In [6]: pandas.__version__
Out[6]: '0.20.2'
```
I think that dask __version__ might be misreporting, pretty sure I'm running on master.

**Comment by mrocklin** [comment]
In this particular case df is probably still in memory as `_6` because of IPython.

**Comment by jcrist** [comment]
> In this particular case df is probably still in memory as _6 because of IPython.

I think that only happens if you ever repr it as output (which didn't happen here). The `In` dictionary is just a mapping to strings.

**Comment by martindurant** [comment]
Python will not necessarily release memory to the OS (this was said above). To test if you are leaking references or need to call gc directly, I wold recommend making a *much* bigger dataframe. I am finding that `del df` is indeed releasing memory, at least sometimes.
For the brave, [objgraph](http://mg.pov.lt/objgraph/#memory-leak-example) may make @pitrou 's analysis above a litle easier.

**Comment by pitrou** [comment]
The real test would be whether memory keeps increasing if you create such dataframes in a loop, instead of reaching a plateau and staying approximately there (or any other upper-bounded memory consumption pattern, such as a sawtooth).

**Comment by shughes-uk** [comment]
Looking at it further it looks like it does plateau, it just plateaus out above my total system memory. I never really had to deal with it before but I was using a new system with a very high core count that I ran a huge amount of workers on, so collectively they would exhaust all the system memory. This can be reclosed now.

**Comment by pitrou** [comment]
The Pandas reference cycle issue will be fixed in the next Pandas release, version 0.22: https://github.com/pandas-dev/pandas/pull/17956
